### PR TITLE
[SPARK-24574][SQL] array_contains, array_position, array_remove and element_at functions deal with Column type

### PR DESCRIPTION
**Pull request by chongguang**
## What changes were proposed in this pull request?

For the function ```def array_contains(column: Column, value: Any): Column ``` , if we pass the `value` parameter as a Column type, it will yield a runtime exception.

This PR proposes a pattern matching to detect if `value` is of type Column. If yes, it will use the .expr of the column, otherwise it will work as it used to.

Same thing for ```array_position, array_remove and element_at``` functions

## How was this patch tested?

Unit test modified to cover this code change.

Ping @ueshin

### CODE REVIEW

File: sql/core/src/main/scala/org/apache/spark/sql/functions.scala
Comment (maropu) [inline sql/core/src/main/scala/org/apache/spark/sql/functions.scala]: We need to update this comment? I think `content of value` is a little ambiguous.

**Comment by maropu** [comment]
BTW, I found the other 3 similar issues there;
```
scala> Seq((Seq(1, 2, 3), 2)).toDF("a", "b").selectExpr("array_position(a, b)").show
+--------------------+
|array_position(a, b)|
+--------------------+
|                   2|
+--------------------+

scala> Seq((Seq(1, 2, 3), 2)).toDF("a", "b").selectExpr("element_at(a, b)").show
+----------------+
|element_at(a, b)|
+----------------+
|               2|
+----------------+

scala> Seq((Seq(1, 2, 3), 2)).toDF("a", "b").selectExpr("array_remove(a, b)").show
+------------------+
|array_remove(a, b)|
+------------------+
|            [1, 3]|
+------------------+
```
I think this is a tiny fix, so IMHO this pr might need to address all the issues here. cc: @ueshin

**Comment by HyukjinKwon** [comment]
ok to test

**Comment by maropu** [comment]
retest this please

File: sql/core/src/main/scala/org/apache/spark/sql/functions.scala
Comment (chongguang) [reply sql/core/src/main/scala/org/apache/spark/sql/functions.scala]: Hey, thanks for the message. Do you want me to change the comment back? I see that you have started the test, is it too late?

Comment (maropu) [reply sql/core/src/main/scala/org/apache/spark/sql/functions.scala]: you can fix now

File: sql/core/src/main/scala/org/apache/spark/sql/functions.scala
Comment (viirya) [inline sql/core/src/main/scala/org/apache/spark/sql/functions.scala]: Other similar expressions are `ArrayPosition`, `ElementAt`, `ArrayRemove`.

Comment (viirya) [reply sql/core/src/main/scala/org/apache/spark/sql/functions.scala]: Oh, I just found @maropu has pointed it out in https://github.com/apache/spark/pull/21581#issuecomment-397928010.

File: sql/core/src/main/scala/org/apache/spark/sql/functions.scala
Comment (ueshin) [inline sql/core/src/main/scala/org/apache/spark/sql/functions.scala]: On second thoughts, we should use `lit()`, e.g., `ArrayContains(column.expr, lit(value).expr)`? WDYT? @viirya @maropu @HyukjinKwon

File: sql/core/src/test/scala/org/apache/spark/sql/DataFrameFunctionsSuite.scala
Comment (ueshin) [inline sql/core/src/test/scala/org/apache/spark/sql/DataFrameFunctionsSuite.scala]: Can you add another test to use `selectExpr`, e.g., `df.selectExpr("array_contains(a, c)")`?

**Review by ueshin — COMMENTED** [review]
I'm sorry for the delay.
I left some comments. Thanks!

File: sql/core/src/main/scala/org/apache/spark/sql/functions.scala
Comment (HyukjinKwon) [reply sql/core/src/main/scala/org/apache/spark/sql/functions.scala]: Yup, that's what I was thinking too from my glance.

Comment (viirya) [reply sql/core/src/main/scala/org/apache/spark/sql/functions.scala]: Yes, I think so.

Comment (maropu) [reply sql/core/src/main/scala/org/apache/spark/sql/functions.scala]: +1

Comment (ueshin) [inline sql/core/src/main/scala/org/apache/spark/sql/functions.scala]: @chongguang I mean, use just `ArrayContains(column.expr, lit(value).expr)` instead of `value match { ...`. The `lit()` should handle literals and columns well.

**Comment by chongguang** [comment]
@ueshin Done! :)

**Comment by HyukjinKwon** [comment]
you could just push the newer changes. that will retrigger the tests.

**Comment by ueshin** [comment]
Jenkins, retest this please.

**Comment by ueshin** [comment]
LGTM pending tests.

**Review by HyukjinKwon — APPROVED** [review]
LGTM too

**Comment by ueshin** [comment]
@chongguang I think this pr is ready to merge, so I tried, but seems like the commits in this pr aren't connected with your GitHub account. If you want to connect the merge commit to your account, could you let me know the email address connected to your account? Thanks!

**Comment by HyukjinKwon** [comment]
oh haha FYI that works after it's merged if @chongguang link the email into his Github profile. I asked the same thing in databricks/spark-xml before :)

**Comment by ueshin** [comment]
But the emails of commits in this pr seem not valid, just for the local computer.

**Comment by HyukjinKwon** [comment]
Oh, I see.

**Comment by chongguang** [comment]
Hey @ueshin I just updated the email address linked to my github account, it is now lcg31439@gmail.com

Thanks

**Comment by HyukjinKwon** [comment]
Merged to master.

**Pull request closed without merging**
